### PR TITLE
Add new resource and data source for Private Link Service

### PR DESCRIPTION
**Pull request by neil-yechenwei**

### CODE REVIEW

File: azurerm/data_source_private_link_service.go [hunk 1]
```diff
@@ -0,0 +1,474 @@
...
+										Type:     schema.TypeList,
+										Computed: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+									},
+								},
+							},
+						},
+						"public_ip_prefix": {
```
Comment (WodansSon) [inline azurerm/data_source_private_link_service.go]: could we flatten this to be just `public_ip_prefix_id`?

File: azurerm/data_source_private_link_service.go [hunk 2]
```diff
@@ -0,0 +1,474 @@
...
+										Computed: true,
+									},
+								},
+							},
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"virtual_machine": {
```
Comment (WodansSon) [inline azurerm/data_source_private_link_service.go]: could we flatten this to be just `virtual_machine_id`?

File: azurerm/resource_arm_private_link_service.go [hunk 3]
```diff
@@ -0,0 +1,1006 @@
...
+			},
+
+			"ip_configurations": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go:79]: Can we get some validation in here?

File: azurerm/resource_arm_private_link_service.go [hunk 4]
```diff
@@ -0,0 +1,1006 @@
...
+						},
+						"private_ip_address": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: same here, some validation would be nice. :)

File: azurerm/resource_arm_private_link_service.go [hunk 5]
```diff
@@ -0,0 +1,1006 @@
...
+						},
+						"subnet": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
+										Type:     schema.TypeString,
+										Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: can we add some validation here?

File: azurerm/resource_arm_private_link_service.go [hunk 6]
```diff
@@ -0,0 +1,1006 @@
...
+									},
+									"name": {
+										Type:     schema.TypeString,
+										Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: and here... can we add some validation?

File: azurerm/resource_arm_private_link_service.go [hunk 7]
```diff
@@ -0,0 +1,1006 @@
...
+					},
+				},
+			},
+
+			"load_balancer_frontend_ip_configurations": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go:114]: is `id` really an optional parameter? also can we add some validation?

File: azurerm/resource_arm_private_link_service.go [hunk 8]
```diff
@@ -0,0 +1,1006 @@
...
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"name": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: yep you guessed it... can we add some validation?

File: azurerm/resource_arm_private_link_service.go [hunk 9]
```diff
@@ -0,0 +1,1006 @@
...
+						},
+						"private_ip_address": {
+							Type:     schema.TypeString,
+							Optional: true,
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: and here... pretty much anywhere there is a string input we need to add some validation, if there isn't a specific rule or set of rules for the attribute at the very least we need to add a not empty check.

File: azurerm/resource_arm_private_link_service.go [hunk 10]
```diff
@@ -0,0 +1,1006 @@
...
+						},
+						"public_ip_address": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: again here, is `id` really a optional attribute?

File: azurerm/resource_arm_private_link_service.go [hunk 11]
```diff
@@ -0,0 +1,1006 @@
...
+										Type:     schema.TypeList,
+										Optional: true,
+										Elem: &schema.Schema{
+											Type: schema.TypeString,
+										},
+									},
+								},
+							},
+						},
+						"public_ip_prefix": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: can we flatten this to `public_ip_prefix_id` , add validation, and verify that `id` is an optional attribute.

File: azurerm/resource_arm_private_link_service.go [hunk 12]
```diff
@@ -0,0 +1,1006 @@
...
+					},
+				},
+			},
+
+			"private_endpoint_connections": {
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go:129]: there are a lot of `id`s that are optional... can we verify that they are in fact optional. Thanks.

File: azurerm/resource_arm_private_link_service.go [hunk 13]
```diff
@@ -0,0 +1,1006 @@
...
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"private_endpoint": {
+							Type:     schema.TypeList,
+							Optional: true,
+							MaxItems: 1,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"id": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go:146]: same thing here, optional `id`?

File: azurerm/resource_arm_private_link_service.go [hunk 14]
```diff
@@ -0,0 +1,1006 @@
...
+										Optional: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+
+			"tags": tags.Schema(),
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: can we move this to the bottom of the schema definition?

File: azurerm/resource_arm_private_link_service.go [hunk 15]
```diff
@@ -0,0 +1,1006 @@
...
+										Optional: true,
+									},
+								},
+							},
+						},
+						"type": {
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"virtual_machine": {
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go]: maybe we can flatten this to just `virtual_machine_id`?

File: azurerm/resource_arm_private_link_service.go [hunk 16]
```diff
@@ -0,0 +1,1006 @@
...
+				},
+			},
+
+			"type": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service.go:228]: It looks like all of the attributes defined in this schema are marked as `optional` can we double check to see if that is true. I find that hard to believe, but they might be... thanks.

File: azurerm/resource_arm_private_link_service_test.go [hunk 17]
```diff
@@ -0,0 +1,309 @@
...
+					resource.TestCheckResourceAttr(resourceName, "ip_configurations.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "ip_configurations.0.subnet.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "ip_configurations.0.private_ip_address", "10.5.1.17"),
+					resource.TestCheckResourceAttr(resourceName, "ip_configurations.0.private_ip_address_version", "IPv4"),
+					resource.TestCheckResourceAttr(resourceName, "ip_configurations.0.private_ip_allocation_method", "Static"),
+					resource.TestCheckResourceAttr(resourceName, "load_balancer_frontend_ip_configurations.#", "1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.%", "1"),
+					resource.TestCheckResourceAttr(resourceName, "tags.env", "test"),
+				),
+			},
```
Comment (WodansSon) [inline azurerm/resource_arm_private_link_service_test.go:116]: Can we add a step the updates it back to the basic configuration and add a import verification step too?

File: examples/virtual-networks/private-link-service/variables.tf [hunk 18]
```diff
@@ -0,0 +1,7 @@
+variable "resource_group_name" {
+  description = "The name of the resource group the Private Link Service is located in."
+}
+
+variable "location" {
+  description = "The Azure location where all resources in this example should be created."
+}
```
Comment (WodansSon) [inline examples/virtual-networks/private-link-service/variables.tf:7]: Would it be possible to add a newline here?

**Review by WodansSon — CHANGES_REQUESTED** [review]
@neil-yechenwei , thanks for the PR... it mostly LGTM but I left some minor comments in the PR. 👍

File: examples/virtual-networks/private-link-service/variables.tf (hunk 18 above)
Comment (neil-yechenwei) [reply examples/virtual-networks/private-link-service/variables.tf:7]: Updated

File: azurerm/resource_arm_private_link_service.go (hunk 14 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Updated

File: azurerm/data_source_private_link_service.go (hunk 1 above)
Comment (neil-yechenwei) [reply azurerm/data_source_private_link_service.go]: Removed unnecessary field.

File: azurerm/data_source_private_link_service.go (hunk 2 above)
Comment (neil-yechenwei) [reply azurerm/data_source_private_link_service.go]: Removed unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 3 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go:79]: Updated code.

File: azurerm/resource_arm_private_link_service.go (hunk 4 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Updated

File: azurerm/resource_arm_private_link_service.go (hunk 5 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Updated

File: azurerm/resource_arm_private_link_service.go (hunk 6 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Removed this unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 7 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go:114]: Yes. It's optional as input value after checked.

File: azurerm/resource_arm_private_link_service.go (hunk 8 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Removed this unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 9 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Removed this unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 10 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Yes. It's read-only field and unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 11 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Removed this unnecessary field.

File: azurerm/resource_arm_private_link_service.go (hunk 12 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go:129]: I've updated all Ids after double checked. Thanks.

File: azurerm/resource_arm_private_link_service.go (hunk 13 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go:146]: Yes. It's optional after checked according by azure rest api specs.

File: azurerm/resource_arm_private_link_service.go (hunk 15 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go]: Removed unnecessary field

File: azurerm/resource_arm_private_link_service.go (hunk 16 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service.go:228]: I've updated code after double checked. Thanks.

File: azurerm/resource_arm_private_link_service_test.go (hunk 17 above)
Comment (neil-yechenwei) [reply azurerm/resource_arm_private_link_service_test.go:116]: Updated code.

**Comment by neil-yechenwei** [comment]
@jeffreyCline , Thanks for your comments. I've updated code. Please help to take a look. Thanks.

File: azurerm/resource_arm_private_link_service.go [hunk 19]
```diff
@@ -0,0 +1,652 @@
...
+
+func resourceArmPrivateLinkServiceCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).network.PrivateLinkServiceClient
+	ctx := meta.(*ArmClient).StopContext
+
+	name := d.Get("name").(string)
+	resourceGroup := d.Get("resource_group_name").(string)
+
+	if features.ShouldResourcesBeImported() && d.IsNewResource() {
```
Comment (magodo) [inline azurerm/resource_arm_private_link_service.go:237]: Would you please explain the error handling logic below? This looks a bit confusing.

**Pull request closed without merging**